rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,850 @@
+
+              case Leaf(leafPrefix, value) =>
+                if (leafPrefix == prefix.tail) value.some.asRight[Params].pure // Happy end
+                else Option.empty[ByteVector].asRight[Params].pure             // Not found
+
+              case NodePtr(ptrPrefix, ptr) =>
+                val (_, prefixRest, ptrPrefixRest) = commonPrefix(prefix.tail, ptrPrefix)
+                if (ptrPrefixRest.isEmpty) loadNode(ptr).map(n => (n, prefixRest).asLeft) // Deeper
+                else Option.empty[ByteVector].asRight[Params].pure                        // Not found
+            }
+        }
+      (startNode, startPrefix).tailRecM(loop)
+    }
+
+    /**
+      * Create node from [[Item]].
+      *
+      * If item is NodePtr and prefix is empty - load child node
+      */
+    private def createNodeFromItem(item: Item): F[Node] =
+      item match {
+        case EmptyItem => emptyNode.pure
+        case Leaf(leafPrefix, leafValue) =>
+          assert(
+            leafPrefix.nonEmpty,
+            "Impossible to create a node. LeafPrefix should be non empty."
+          )
+          emptyNode.updated(byteToInt(leafPrefix.head), Leaf(leafPrefix.tail, leafValue)).pure
+        case NodePtr(nodePtrPrefix, ptr) =>
+          if (nodePtrPrefix.isEmpty) loadNode(ptr)
+          else
+            emptyNode.updated(byteToInt(nodePtrPrefix.head), NodePtr(nodePtrPrefix.tail, ptr)).pure
+      }
+
+    /**
+      * Optimize and save Node, create item from this Node
+      */
+    private def saveNodeAndCreateItem(
+        node: Node,
+        prefix: ByteVector,
+        compaction: Boolean = true
+    ): Item =
+      if (compaction) {
+        val nonEmptyItems = node.iterator.filter(_ != EmptyItem).take(2).toList
+        nonEmptyItems.size match {
+          case 0 => EmptyItem // All items are empty.
+          case 1 => // Only one item is not empty - merge child and parent nodes.
+            val idxItem = node.indexOf(nonEmptyItems.head)
+            nonEmptyItems.head match {
+              case EmptyItem => EmptyItem
+              case Leaf(leafPrefix, value) =>
+                Leaf(prefix ++ ByteVector(idxItem) ++ leafPrefix, value)
+              case NodePtr(nodePtrPrefix, ptr) =>
+                NodePtr(prefix ++ ByteVector(idxItem) ++ nodePtrPrefix, ptr)
+            }
+          case 2 => // 2 or more items are not empty.
+            NodePtr(prefix, saveNode(node))
+        }
+      } else NodePtr(prefix, saveNode(node))
+
+    /**
+      * Save new leaf value to this part of tree (start from curItems).
+      * Rehash and save all depend node to [[cacheW]].
+      *
+      * If exist leaf with same prefix but different value - update leaf value.
+      * If exist leaf with same prefix and same value - return [[None]].
+      * @return Updated current item.
+      */
+    def update(
+        curItem: Item,
+        insPrefix: ByteVector,
+        insValue: ByteVector
+    ): F[Option[Item]] =
```

Comment:
Here is similar situation as with commit. Some parts have _calculation_ which is not part of `F` and will be executed immediately when _update_ function is invoked, even before `F` execution started or not executed at all.

As a good measure when _pure_ can be used instead of _suspend_ is, use _pure_ only for values that are already calculated so no additional function call is necessary. In other cases where more work is needed use _suspend_ to move this work as part of `F` execution.

Reply:
@tgrospic, I guess it is relaited to your example in [rchain-dev-playground](https://github.com/tgrospic/rchain-dev-playground/blob/master/src/main/scala/dev_00_intro/_10_F_context.scala) repo.